ogr/services/forgejo/pull_request.py

```diff
@@ -27,15 +153,197 @@ def create(
+            repo=target_project.repo,
+            base=target_branch,
+            body=body,
+            head=source_branch,
+            title=title,
+        )
+        logger.info(f"PR {pr.id} created.")
+
+        return ForgejoPullRequest(pr, target_project)
 
     @staticmethod
     def get(project: "forgejo.ForgejoProject", pr_id: int) -> "PullRequest":
-        raise NotImplementedError("TBD")
+        try:
+            raw_pr = project.api.repo_get_pull_request(
+                owner=project.namespace,
+                repo=project.repo,
+                index=pr_id,
+            )
+        except NotFoundError as ex:
+            raise ForgejoAPIException(f"No pull request with id {pr_id} found.") from ex
+        return ForgejoPullRequest(raw_pr, project)
 
     @staticmethod
     def get_list(
         project: "forgejo.ForgejoProject",
         status: PRStatus = PRStatus.open,
     ) -> list["PullRequest"]:
-        raise NotImplementedError("TBD")
+        prs = project.api.repo_list_pull_requests(
+            owner=project.namespace,
+            repo=project.repo,
+            # Forgejo has just open/closed/all
+            state=status.name if status != PRStatus.merged else "closed",
+        )
+        return [ForgejoPullRequest(pr, project) for pr in prs]
+
+    def update_info(
+        self,
+        title: Optional[str] = None,
+        description: Optional[str] = None,
+    ) -> "PullRequest":
+        try:
+            data = {"title": title if title else self.title}
+
+            if description is not None:
+                data["body"] = description
+
+            updated_pr = self._target_project.api.repo_edit_pull_request(
+                owner=self.target_project.namespace,
+                repo=self.target_project.repo,
+                index=self.id,
+                **data,
+            )
+
+            self._raw_pr = updated_pr
+            return self
+        except Exception as ex:
+            raise ForgejoAPIException(
+                f"There was an error while updating Forgejo PR: {ex}",
+            ) from ex
+
+    def close(self) -> "PullRequest":
+        self._raw_pr = self._target_project.api.repo_edit_pull_request(
+            owner=self.target_project.namespace,
+            repo=self.target_project.repo,
+            index=self.id,
+            state="closed",
+        )
+        return self
+
+    def merge(self) -> "PullRequest":
+        self._target_project.api.repo_merge_pull_request(
+            owner=self.target_project.namespace,
+            repo=self.target_project.repo,
+            index=self.id,
+            # options: merge, rebase, rebase-merge, squash, fast-forward-only, manually-merged
+            do="merge",
+        )
+        return self.get(self._target_project, self.id)
+
+    def add_label(self, *labels: str) -> None:
+        raise NotImplementedError("Not possible via Forgejo API.")
+
+    def get_all_commits(self) -> list[str]:
```

Comment:
I'd expect pagination + iterable here too

Reply:
Makes sense to me, though you still need to paginate :/

Reply:
PTAL